src/snowflake/snowpark/_internal/code_generation.py

```diff
@@ -0,0 +1,594 @@
+        top_level_class = base_class_name.split(".")[0]
+        if base_class_name == "object":
+            continue
+        elif base_class.__module__ == func_module_name:
+            # if base class is from the same module, we need to parse the class as well as generate code for the class
+            # False in the tuple means this is not a nested class
+            inferred_classes.append((base_class, False))
+        else:
+            # if base class is from another module, we need to import it
+            ref_objects[top_level_class] = func.__globals__[top_level_class]
+
+    # __dict__ contains function, classmethod, classes attributes within a given class which
+    # needs to be further analyzed
+    for v in dict(cls.__dict__).values():
+        if inspect.isclass(v):
+            top_level_cls_name = v.__qualname__.split(".")[0]
+            if v.__module__ == func_module_name:
+                # v is a class defined in the same module as UDF func's, need to dynamically parse the class
+                # if v is class defined in cls, then we should not re-generate code for the nested class
+                inferred_classes.append(
+                    (v, v.__qualname__.startswith(top_level_cls_name))
+                )
+            else:
+                # v is a class defined in another module, import the top level class from another module
+                ref_objects[top_level_cls_name] = v
+        elif inspect.isfunction(v) or isinstance(v, classmethod):
+            # v is a function/classmethod, get the references objects of the Function object
+            get_func_references(
+                v if not isinstance(v, classmethod) else v.__func__, ref_objects
+            )
+        else:
+            # cls.__dict__ would also return __module__, __doc__, __weakref__ which are not required
+            # for code generation, however, class variables is also included in __dict__, we don't do value evaluation
+            # for them in the current implementation (e.g. the declaration of class variables is assigned the
+            # result of function call). But we shall introduce an argument to control the behavior,
+            # check JIRA SNOW-649884
+            pass
+
+    # recursively handling inferred classes that should be analyzed dynamically
+    for inferred_class, is_nested_class in inferred_classes:
+        get_class_references(
+            inferred_class,
+            func,
+            ref_objects,
+            classes_to_generate,
+            generate_code_for_class=not is_nested_class,
+        )
+
+
+def extract_func_global_refs(code: CodeType) -> Dict[str, None]:
+    # inspired by cloudpickle to recursively extract all the global references used by the target func's code object
+    co_names = code.co_names
+    out_names = {}
+    for instr in dis.get_instructions(code):
+        op = instr.opcode
+        if op in GLOBAL_OPS:
+            out_names[co_names[instr.arg]] = None
```

Comment:
Is `instr.arg` guaranteed to be part of `co_names`? Also, why do we set it to `None`?

Reply:
yes, instr.arg is 1-1 mapping parameter position info for the co_names.

as for none in dict, it is copied from cloudpickle, cloudpickle uses dict for deterministic ordering of pickled bytes. we don't need it. I've changed to use set.